resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/tpr/TPRCrudCallsImpl.scala

```diff
@@ -0,0 +1,320 @@
+}
+
+private[spark] case class Metadata(name: String,
+    uid: Option[String] = None,
+    labels: Option[Map[String, String]] = None,
+    annotations: Option[Map[String, String]] = None)
+
+private[spark] case class SparkJobState(apiVersion: String,
+    kind: String,
+    metadata: Metadata,
+    status: Map[String, Any])
+
+private[spark] case class WatchObject(`type`: String, `object`: SparkJobState)
+
+/**
+ * Prereq - This assumes the kubernetes API has been extended using a TPR of name: Spark-job
+ * See conf/kubernetes-custom-resource.yaml for a model
+ *
+ * CRUD + Watch Operations on SparkJob Resource
+ *
+ * This class contains all CRUD+Watch implementations performed
+ * on the SparkJob Resource used to expose the state of a spark job
+ * in kubernetes (visible via kubectl or through the k8s dashboard).
+ */
+private[spark] class TPRCrudCallsImpl(k8sClient: KubernetesClient,
+    kubeToken: Option[String] = None) extends Logging with TPRCrudCalls {
+  private[spark] val kubeMaster: String = k8sClient.getMasterUrl().toString
+  private val httpClient: OkHttpClient =
+    extractHttpClientFromK8sClient(k8sClient.asInstanceOf[BaseClient])
+  private val namespace: String = k8sClient.getNamespace
+  private var watchSource: BufferedSource = _
+  private lazy val buffer = new Buffer()
+
+  private implicit val formats: Formats = DefaultFormats + JobStateSerDe
+  private implicit val ec: ThreadPoolExecutor = ThreadUtils
+    .newDaemonCachedThreadPool("tpr-watcher-pool")
+
+  override def createJobObject(name: String, keyValuePairs: Map[String, Any]): Unit = {
+    val resourceObject =
+      SparkJobState(s"$TPR_API_GROUP/$TPR_API_VERSION", TPR_KIND, Metadata(name), keyValuePairs)
+    val payload = parse(write(resourceObject))
+
+    val requestBody = RequestBody
+      .create(MediaType.parse("application/json"), compact(render(payload)))
+
+    val requestSegments = Seq(
+      "apis", TPR_API_GROUP, TPR_API_VERSION, "namespaces", namespace, "sparkjobs")
+    val url = generateHttpUrl(requestSegments)
+
+    val request = completeRequest(new Request.Builder()
+      .post(requestBody)
+      .url(url))
+
+    logDebug(s"Create SparkJobResource Request: $request")
+    val response = httpClient.newCall(request).execute()
+    completeRequestWithExceptionIfNotSuccessful(
+      "post",
+      response,
+      Option(Seq(name, response.toString, compact(render(payload)))))
+
+    response.body().close()
+    logDebug(s"Successfully posted resource $name: " +
+      s"${pretty(render(parse(write(resourceObject))))}")
+  }
+
+  override def deleteJobObject(tprObjectName: String): Unit = {
+    val requestSegments = Seq(
+      "apis", TPR_API_GROUP, TPR_API_VERSION, "namespaces", namespace, "sparkjobs", tprObjectName)
+    val url = generateHttpUrl(requestSegments)
+
+    val request = completeRequest(new Request.Builder()
+      .delete()
+      .url(url))
+
+    logDebug(s"Delete Request: $request")
```

Comment:
does this log credentials used in the request out to the log?  definitely want to avoid logging secrets to files